src/python/pants/backend/jvm/tasks/analysis_extraction.py

```diff
@@ -47,106 +46,89 @@ def _create_products_if_should_run(self):
+      with self.invalidated(targets,
+                            fingerprint_strategy=fingerprint_strategy,
+                            invalidate_dependents=True) as invalidation_check:
+        for vt in invalidation_check.all_vts:
+          # class paths for the target we are computing deps for
+          target_cps = [cp[1] for cp in classpath_product.get_for_target(vt.target)]
+
+          jdeps_output_json = self._jdeps_output_json(vt)
+          if not vt.valid:
+            self._run_jdeps_analysis(vt.target, target_cps, deps_classpath_by_alias, jdeps_output_json)
+          self._register_products(vt.target,
+                                  jdeps_output_json,
+                                  product_deps_by_target)
+
+  @memoized_property
+  def _jdeps_summary_line_regex(self):
+    return re.compile(r"^\S+\s->\s(\S+)$")
+
+  def _run_jdeps_analysis(self, target, target_cps, deps_classpath_by_alias, jdeps_output_json):
+    with open(jdeps_output_json, 'w') as f:
+      if target_cps:
+        # TODO should we find an abs path to jdeps in a better way? a jdk path?
+        cmd = [
+            "jdeps", "-summary",
+            '-classpath', ":".join(cp for cp in deps_classpath_by_alias.keys()),
+          ] + target_cps
+        jdeps_output = io.StringIO(subprocess.run(cmd, stdout=subprocess.PIPE).stdout.decode('utf-8'))
+        deps_classpaths = set()
+        for line in jdeps_output:
+          match = self._jdeps_summary_line_regex.fullmatch(line.strip()).group(1)
+          deps_classpaths.add(deps_classpath_by_alias.get(match, match))
+
+      else:
+        deps_classpaths = []
+      json.dump(list(deps_classpaths), f)
 
   def _register_products(self,
                          target,
-                         target_cp_entry,
-                         summary_json_file,
-                         classes_by_source,
-                         product_deps_by_src):
-    summary_json = self._parse_summary_json(summary_json_file)
-
-    # Register a mapping between sources and classfiles (if requested).
-    if classes_by_source is not None:
-      buildroot = get_buildroot()
-      for abs_src, classes in summary_json['products'].items():
-        source = os.path.relpath(os.path.normpath(abs_src), buildroot)
-        classes = [os.path.normpath(c) for c in classes]
-        classes_by_source[source].add_abs_paths(target_cp_entry, classes)
-
-    # Register classfile product dependencies (if requested).
-    if product_deps_by_src is not None:
-      product_deps_by_src[target] = {
-          os.path.normpath(src): [os.path.normpath(dep) for dep in deps]
-          for src, deps in summary_json['dependencies'].items()
-        }
-
-  def _parse_summary_json(self, summary_json_file):
-    with open(summary_json_file, 'r') as f:
-      return json.load(f, encoding='utf-8')
+                         jdeps_output_json,
+                         product_deps_by_target):
+    if target not in product_deps_by_target:
```

Comment:
It would be odd if a target was already in this list... safe to ignore that case, and assume that this task is the only thing writing this product.